internal/datastore/crdb/watch.go

```diff
@@ -195,7 +206,192 @@ func (cds *crdbDatastore) Watch(ctx context.Context, afterRevision datastore.Rev
+					IsCheckpoint: true,
+				}
+				continue
+			}
+
+			tableName := string(tableNameBytes)
+
+			var pkValues []string
+			if err := json.Unmarshal(primaryKeyValuesJSON, &pkValues); err != nil {
+				errs <- err
+				return
+			}
+
+			if len(pkValues) != 1 {
+				errs <- spiceerrors.MustBugf("expected a single definition name for the primary key in schema change feed. found: %s", string(primaryKeyValuesJSON))
+				return
+			}
+
+			definitionName := pkValues[0]
+
+			deletedNamespaces := make([]string, 0, 1)
+			deletedCaveats := make([]string, 0, 1)
+			changedDefinitions := make([]datastore.SchemaDefinition, 0, 1)
+
+			switch tableName {
+			case tableNamespace:
+				if details.After != nil && details.After.SerializedNamespaceConfig != "" {
+					namespaceDef := &core.NamespaceDefinition{}
+					defBytes, err := hex.DecodeString(details.After.SerializedNamespaceConfig[2:]) // drop the \x
+					if err != nil {
+						errs <- fmt.Errorf("could not decode namespace definition: %w", err)
+						return
+					}
+
+					if err := namespaceDef.UnmarshalVT(defBytes); err != nil {
+						errs <- fmt.Errorf("could not unmarshal namespace definition: %w", err)
+						return
+					}
+					changedDefinitions = append(changedDefinitions, namespaceDef)
+				} else {
+					deletedNamespaces = append(deletedNamespaces, definitionName)
+				}
+
+			case tableCaveat:
+				if details.After != nil && details.After.SerializedCaveatDefinition != "" {
+					caveatDef := &core.CaveatDefinition{}
+					defBytes, err := hex.DecodeString(details.After.SerializedCaveatDefinition[2:]) // drop the \x
+					if err != nil {
+						errs <- fmt.Errorf("could not decode caveat definition: %w", err)
+						return
+					}
+
+					if err := caveatDef.UnmarshalVT(defBytes); err != nil {
+						errs <- fmt.Errorf("could not unmarshal caveat definition: %w", err)
+						return
+					}
+					changedDefinitions = append(changedDefinitions, caveatDef)
+				} else {
+					deletedCaveats = append(deletedCaveats, definitionName)
+				}
+
+			default:
+				errs <- spiceerrors.MustBugf("unknown table in schema change feed: %s", tableName)
+				return
+			}
+
+			revision, err := cds.RevisionFromString(details.Updated)
+			if err != nil {
+				errs <- fmt.Errorf("malformed resolved timestamp: %w", err)
+				return
+			}
+
+			updates <- &datastore.SchemaState{
+				Revision:           revision,
+				ChangedDefinitions: changedDefinitions,
+				DeletedNamespaces:  deletedNamespaces,
+				DeletedCaveats:     deletedCaveats,
+				IsCheckpoint:       false,
+			}
+		}
+
+		if changes.Err() != nil {
+			if errors.Is(ctx.Err(), context.Canceled) {
+				closeCtx, closeCancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer closeCancel()
+				if err := conn.Close(closeCtx); err != nil {
+					errs <- wrapErrorIfRetryable(ctx, err)
+				}
+				errs <- datastore.NewWatchCanceledErr()
+			} else {
+				errs <- wrapErrorIfRetryable(ctx, changes.Err())
```

Comment:
Instead of bubbling up through the interface, why not just reset the connection in `WatchSchema` here?

Then you don't even need to export the Retryable error, and the caching proxy doesn't have to worry about restarting the watch? 

If you need to reset the connection for a retryable error, the pgconn you created just for the watch needs to be closed too (I don't see that happening here?)

Reply:
Partially because it breaks the ability for the cache to know to "reset" itself; it may have missed events between the calls and so it should do a full reset of the in-memory cache to ensure it has a complete picture of the schema changes